sklearn/utils/stats.py

```diff
@@ -101,22 +114,55 @@ def _weighted_percentile(array, sample_weight, percentile_rank=50, xp=None):
             for feature_idx in range(weight_cdf.shape[0])
         ],
     )
-    # In rare cases, `percentile_indices` equals to `sorted_idx.shape[0]`
+    # `percentile_indices` may be equal to `sorted_idx.shape[0]` due to floating
+    # point error (see #11813)
     max_idx = sorted_idx.shape[0] - 1
     percentile_indices = xp.clip(percentile_indices, 0, max_idx)
 
     col_indices = xp.arange(array.shape[1], device=device)
     percentile_in_sorted = sorted_idx[percentile_indices, col_indices]
 
-    result = array[percentile_in_sorted, col_indices]
+    if average:
+        # From Hyndman and Fan (1996), `fraction_above` is `g`
+        fraction_above = (
+            weight_cdf[col_indices, percentile_indices] - adjusted_percentile_rank
+        )
+        # Alternatively, could use
+        # `is_exact_percentile = fraction_above <= xp.finfo(floating_dtype).eps`
+        # but that seems harder to read
+        is_fraction_above = fraction_above > xp.finfo(floating_dtype).eps
+        percentile_plus_one_in_sorted = sorted_idx[percentile_indices + 1, col_indices]
+        # Handle case when when next index ('plus one') has sample weight of 0
+        if xp.any(sample_weight[percentile_plus_one_in_sorted, col_indices] == 0):
+            for idx in col_indices:
+                if (
+                    sample_weight[percentile_plus_one_in_sorted[idx], col_indices[idx]]
+                    == 0
+                ):
+                    cdf_val = weight_cdf[col_indices[idx], percentile_indices[idx]]
+                    # Search for next index where `weighted_cdf` is greater
+                    next_cdf_index = xp.searchsorted(
+                        weight_cdf[idx, ...], cdf_val, side="right"
+                    )
+                    # To account for trailing 0 sample weights, use original
+                    # `percentile_indices` again
+                    if next_cdf_index >= max_idx:
+                        next_cdf_index = percentile_indices[idx]
+
+                    percentile_plus_one_in_sorted[idx] = sorted_idx[
+                        next_cdf_index, col_indices
+                    ]
+
+        result = xp.where(
+            is_fraction_above,
+            array[percentile_in_sorted, col_indices],
```

Comment:
I initially thought this should be `percentile_plus_one_in_sorted` as from the paper, when g>0, $`\gamma=1`$, but `searchsorted` defaults to left (equals is on the right), whereas the paper defined $`j\le pn\lt j+1`$

Reply:
Maybe you could include the remark about the discrepancy between `searchsorted` and the paper convention as an inline comment in the code.

Reply:
Done, hope it is adequate?